### PR TITLE
milestone #3 -gamelogic

### PR DESCRIPTION
**Pull request by IBUNHABIBU**
added the code for gamelogic branch

### CODE REVIEW

**Comment by Oghenebrume50** [comment]
reviewing......

File: lib/board.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    def replace_var(position,token)
+      @remval.delete(@board[position])
+      @board[position] = token
+    end
+    def rem_nums
+      return @remval
+    end
+    def display
+      (1..9).each do |i|
+        print " | " + @board[i - 1].to_s
```
Comment (Oghenebrume50) [inline lib/board.rb]: alll display logic should be in one class

File: lib/board.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    ]
+    win_combination.each{ |element|
+     bool_value = [@board[element[0]],@board[element[1]],@board[element[2]]].all? { |i| i == "x" } || [@board[element[0]],@board[element[1]],@board[element[2]]].all? { |i| i == "o" }
+     if bool_value
+      @winner = @board[element[0]]
+     end
+      break  if bool_value
+    }
+      return bool_value
+  end
```
Comment (Oghenebrume50) [inline lib/board.rb:16]: This particular method could be done better, how about taking in a symbol and checking instead of repeating the search

**Review by Oghenebrume50 — CHANGES_REQUESTED** [review]
Hi Team good one on this, please fix the outline stuff and submit for re-review.

File: lib/board.rb (hunk 1 above)
Comment (Elukoye) [reply lib/board.rb]: Please clarify what you mean by this
The display function is defined inside the board class.

File: lib/board.rb (hunk 2 above)
Comment (Elukoye) [reply lib/board.rb:16]: We felt that was the best way we could define that method, in a way we understood what was happening and how it worked.

File: bin/main.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
+require_relative'../lib/board.rb'
+class Logic
+    def initialize
+      @player_1 = "player_1"
+      @player_2 = "player_2"
+      @board = Board.new
+    end
+    def game_start
+      puts "Welcome to TIC-TAC-TOE"
```
Comment (bolah2009) [inline bin/main.rb]: #### `REQUIRED`

`main.rb` is the only place where `puts` and `get.chomp` can be used.  :blush: 
Let's refactor the code and do this... :+1: 

> Add executable file named `main.rb` to bin directory. This file should be responsible for user interaction. This is the only one place where `puts` and `get.chomp` can be used.

Let's do this for all other occurrences.

**Review by bolah2009 — CHANGES_REQUESTED** [review]
# Nice Job Team :clap: 

```bash
Bola [x] ,choose one among [9]: 
9
 | x | x | o | 
 | o | o | x | 
 | x | o | x | 
 Damn ,its a draw!!... PLAY AGAIN?
```

The Game works **perfectly** ... :ok_hand: :smile: 

But we seem not to be working the **OOP** way... which is the main objective of this project. Let's do this. I will also be quoting the [pathwright project requirements](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/) :blush: 

- Let's use a different class for different logic. Suggestions: `player`, `game`, `board` classes.
  > Think of how you can encapsulate different parts of the game logic in classes. Do not create one class for everything.
  > Remember about single responsibility rule. Always.

  :bulb: What if `player` takes and displays the players' data, `board` displays board and `game`  checks for winners/draw. Let the classes displaying data return only string so that `bin/main.rb` will be in charge of `print`-ing or `puts`-ing it. :thinking: 
- FIx linter warnings: Is stickler [installed](https://github.com/settings/installations) on this repo? :thinking: Check [here](https://github.com/microverseinc/linters-config/tree/master/ruby#troubleshooting) for troubleshooting.
 Currently, there are **102 offenses detected** with `rubocop -a` locally.
- Resolve merge conflicts
  > Then merge changes from a refreshed `development` branch to `next-milestone` branch and **resolve conflicts**.
- Check out other comments below too :blush: 


Good luck with your next submission. :star2: 

Cheers
[Bola](https://bolabuari.com/)  

###### PS: Note that comments tagged as `OPTIONAL` are not part of the minimum requirements to approve this PR

**Comment by Elukoye** [comment]
Hi bola ,
I thought the project said at least two classes I have a board class and a
game logic class .

**Comment by Elukoye** [comment]
Could we please chat on slack so that you can clarify because so far it
seems every reviewer has something diff to say.

File: bin/main.rb (hunk 3 above)
Comment (Elukoye) [reply bin/main.rb]: If you take a look at the heading of this file it is the min.rb file. the @board is an instance from the class board which is on the board.rb file and has been required into this main file.

**Review by maelfosso — CHANGES_REQUESTED** [review]
Hello @IBUNHABIBU @Elukoye 

Everything is OK !!
The game is perfect :smile: 

The only thing to change is the **project structure**, I mean where should which file be.
**Please move the `Logic` class into the lib/ folder**
Don't forget the rule **no puts and chomps except into `main.rb` file**

Kindly make that change and re-submit a new code review request.
Don't hesitate to message on Slack if you have any question.
Thanks.

**Comment by Elukoye** [comment]
Corrections made, kindly review

File: lib/board.rb [hunk 4]
```diff
@@ -48,4 +48,73 @@ def winner
...
+    @board = Board.new
+  end
+  def game_start
+    puts "Welcome to TIC-TAC-TOE"
+    puts " Player one input name: "
+    @player_1 = gets.chomp.capitalize + " [x]"
+    puts "#{@player_1} is  x"
+    puts "Player two input name: "
+    @player_2 = gets.chomp.capitalize + " [o]"
+   puts "#{@player_2} is  o"
```
Comment (maelfosso) [inline lib/board.rb]: No puts here

File: lib/board.rb [hunk 5]
```diff
@@ -48,4 +48,73 @@ def winner
...
+      position = gets.chomp.to_i
+      while true do
+        if  position.between?(1,9) && !@board.position_taken?(position)
+        position -= 1
+        break
+        else
+        puts "Invalid move or input.\n try again: "
+        position = gets.chomp.to_i
+        end
+      end
```
Comment (maelfosso) [inline lib/board.rb]: No `puts` and `get.chomp` here

**Review by maelfosso — CHANGES_REQUESTED** [review]
Hello coders,

Nice improvement. 
But there are still things to do

1. Could you create a file for each class ? **`OPTIONAL`**  It just make your code more readable
2. Please, correctly indent your code
3. Respect the rules **No puts or chomps inside class**

Thanks

**Comment by Elukoye** [comment]
Kindly review

**Review by Ispirett — APPROVED** [review]
Great job folks, I know you worked hard on this project making changes :100: :1st_place_medal: 
You can merge your code now.

[Optional] Not required 
It will be better to wrap the logic in your main.rb file in a class
On line `55, 66` the winner logic should have its own method.

**Comment by Elukoye** [comment]
Thank you

File: bin/main.rb (hunk 3 above)
Comment (bolah2009) [reply bin/main.rb]: #### Oops! You are right. Seeing that it was the `Logic` class, I mistook it for one of the `lib/*` files.
https://github.com/Elukoye/OOP-Ruby/blob/57fc349d4b5dc3b8367a50af623fb5c8cb22f2b2/bin/main.rb#L2-L9
#### Nice job moving the class away from the `main.rb` file :+1: 
#### Sorry about that... :bowing_man:

**Comment by bolah2009** [comment]
> I thought the project said at least two classes I have a board class and a
game logic class .

#### Yes, that's right. It was a suggestion and I could only see one class (`Board`) in the `lib` directory when I reviewed.

> Could we please chat on slack so that you can clarify because so far it seems every reviewer has something diff to say.

#### Please feel free to chat me up next time. I didn't get the notification of your replies earlier. Also, I would have chat you up both couldn't find @Elukoye on slack.

**Pull request merged**